_posts/2020-08-26-functional-programming-in-java.md

```diff
@@ -0,0 +1,167 @@
+- **Immutability**: Since **avoiding side-effects** is a key paradigm of Functional Programming, states or data structures are kept immutable i.e not modified once they are initialized. If we need to change the state of a variable, we will just create a new variable.
+- **Recursion**: Recursion is the act of calling a function from itself, until a certain condition is satisfied (unless you want to sacrifice your code to an infinite loop).
+
+## Functional Programming in Java 8
+
+Can we reconcile the above concepts in Java? Well, Java is not a "Functional Programming" language. But, leveraging the features of Java 8, we can mimic the major Functional Programming approaches. For example:
+
+Streams and Lambda Expressions: Java 8 provides the Streams API which can take a data structure (Arrays, Lists etc) as the source, and, pass it through a pipeline of functional-styled operations. These operations do not change the source but produces a new output stream (therefore, fulfilling the **no-side effect, immutability** condition of Functional Programming.
+
+For example, we can have a list of Strings. But, we only need the Strings having the length of 4. Using Streams, this becomes easy:
+
+```java
+// Our original list
+List<String> strList = new ArrayList<String>();
+// Let's assume we have added 5 elements in the array: "Golf", "Stoic", "Planet", "Mars", "Moon".
+strList = strList.stream().filter(str -> str.toString().length()==4).collect(Collectors.toList());
+```
+
+The filtered list does not affect the original list, unless we assign it to that (like we're doing here). The `()->` notation in the `filter` method is called Lambda expression.
+
+Functional Interfaces: An interface with only one abstract method is a Functional Interface. Let's see how we can define one:
+
+```java
+@FunctionalInterface
+interface FullName 
+{ 
+    int getFullName(String firstName, String lastName); 
+} 
+```
+
+All we need to do is implement the `getFullName` method. How do we do it? Here, too, lambda expressions come to the rescue. We can directly declare an object of the Functional Interface and assign an implementation of the `getFullName` method:
+
+```java
+FullName fullName = (String firstName, String lastName)->firstName + ' ' + lastName; 
+```
+
+Now, we can call the implemented method on the object. It will return the full name String:
+
+```java
+String resultFullName = fullName.getFullName("John","Doe"); 
+```
+
+That's the easiest way to make a Pure Function. See, the result will always be predictable for the same set of inputs. But, what about Higher-order functions? How would we compose a chain of functions? Well, that's what we'll do by creating a command-line mock banking application. Java 8 provides built-in Functional Interfaces that take functions as input - exactly what we need.
+
+## Bringing It All Together
```

Comment:
There's an asymmetry in the article structure here. First, you explain streams and functional interfaces on their own. Then, you introduce Consumers, Functions, BuFunctions and Predicates in the context of the banking application.

I like the approach of explaining the features along the banking application, but can we do this for all of the functional features?

I would like to see an article structure like this:

* Functional Programming: What, Why, and When?
* Major Concepts of Functional Programming
* Functional  Programming in Java 8 
  * Streams
  * Functional Interfaces
  * Consumers
  * Functions
  * BiFunctions
  * Predicates

Introduce the Banking app in the intro to the "Functional Programming" section and explain each feature directly alongside some code of the banking app (you might need to change the order of the features for it to make sense).

This would give the article a better structure.